### PR TITLE
Fuerte HTTP2 support

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Http2 support in the fuerte client library

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)

#### Related Information

- [ ] There is a GitHub Issue reported by a Community User: 
- [x] There is an internal planning ticket:  https://arangodb.atlassian.net/jira/software/projects/FEAT/boards/7/backlog?selectedIssue=FEAT-205
- [ ] There is a *JIRA Ticket number* (In case a customer was affected / involved): 
- [ ] There is a design document: 

### Testing & Verification

This change is already covered by existing tests, such as shell_client testsuite

- [x] I ensured this code runs with ASan / TSan or other static verification tools

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/8180/

### Documentation

- [x] Added a new section in the *Manual* : https://github.com/arangodb/docs/pull/310

### CODE REVIEW

File: 3rdParty/fuerte/include/fuerte/connection.h [hunk 1]
```diff
@@ -69,14 +69,14 @@ class Connection : public std::enable_shared_from_this<Connection> {
   /// When a response is received or an error occurs, the corresponding
   /// callbackis called. The callback is executed on a specific
   /// IO-Thread for this connection.
-  virtual MessageID sendRequest(std::unique_ptr<Request> r,
-                                RequestCallback cb) = 0;
+  virtual void sendRequest(std::unique_ptr<Request> r,
```
Comment (ObiWahn) [inline 3rdParty/fuerte/include/fuerte/connection.h]: What happens if the CB is never called? Do you still have way to clean up without the id? (Note to myself: check if CB has timeout)

File: 3rdParty/fuerte/src/H2Connection.cpp [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+                Connection::State::Connected);
+  FUERTE_LOG_HTTPTRACE << "startWriting: this=" << this << "\n";
+  bool tmp = _signaledWrite.load();
+  if (!tmp && !_signaledWrite.exchange(true)) {
+    this->_io_context->dispatch([self = Connection::shared_from_this(), this] {
+      _signaledWrite.store(false);
+      // we have been in a race with shutdownConnection()
+      Connection::State state = this->_state.load();
+      if (state != Connection::State::Connected) {
+        //        this->_writing.store(false);
```
Comment (ObiWahn) [inline 3rdParty/fuerte/src/H2Connection.cpp]: can this comment be deleted?

File: 3rdParty/fuerte/include/fuerte/connection.h [hunk 3]
```diff
@@ -69,14 +69,14 @@ class Connection : public std::enable_shared_from_this<Connection> {
...
 
   /// @brief Send a request to the server and return immediately.
   /// When a response is received or an error occurs, the corresponding
   /// callbackis called. The callback is executed on a specific
   /// IO-Thread for this connection.
-  MessageID sendRequest(Request const& r, RequestCallback cb) {
+  void sendRequest(Request const& r, RequestCallback cb) {
     auto copy = std::make_unique<Request>(r);
     return sendRequest(std::move(copy), cb);
   }
```
Comment (jsteemann) [inline 3rdParty/fuerte/include/fuerte/connection.h]: Isn't this just
```
void sendRequest(Request const& r, RequestCallback cb) {
  sendRequest(std::make_unique<Request>(r), cb);
}
```

File: 3rdParty/fuerte/src/AsioSockets.h [hunk 4]
```diff
@@ -58,125 +57,170 @@ void resolveConnect(detail::ConnectionConfiguration const& config,
...
+    } catch (...) {
     }
   }
-  
+
+  template <typename F>
+  void shutdown(F&& cb) {
+    asio_ns::error_code ec;  // prevents exceptions
+    try {
+#ifndef _WIN33
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/AsioSockets.h]:
```suggestion
#ifndef _WIN32
```

File: 3rdParty/fuerte/src/H2Connection.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  void queueHttp2Requests();
+
+  /// actually perform writing
+  void doWrite();
+
+  Stream* findStream(int32_t sid) const;
+
+ private:
+  static constexpr size_t kOutBufferLen = 32 * 1024 * 1024;
+  std::array<uint8_t, kOutBufferLen> _outbuffer;
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H2Connection.h]: This generates a 32 MB buffer per HTTP/2 connection. This seems overly much.
Can't we allocate this buffer on the heap if needed?

File: 3rdParty/fuerte/src/H2Connection.cpp [hunk 6]
```diff
@@ -0,0 +1,808 @@
...
+  populateSettings(iv);
+
+  nghttp2_submit_settings(session, NGHTTP2_FLAG_NONE, iv.data(), iv.size());
+  // increase connection window size up to window_size
+  nghttp2_session_set_local_window_size(session, NGHTTP2_FLAG_NONE, 0, 1 << 30);
+}
+
+std::string makeAuthHeader(fu::detail::ConnectionConfiguration const& config) {
+  std::string auth;
+  // preemtively cache authentication
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H2Connection.cpp]:
```suggestion
  // preemptively cache authentication
```

File: 3rdParty/fuerte/src/H2Connection.cpp [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+  iv[2] = {NGHTTP2_SETTINGS_ENABLE_PUSH, 0};
+}
+
+void submitConnectionPreface(nghttp2_session* session) {
+  std::array<nghttp2_settings_entry, 3> iv;
+  populateSettings(iv);
+
+  nghttp2_submit_settings(session, NGHTTP2_FLAG_NONE, iv.data(), iv.size());
+  // increase connection window size up to window_size
+  nghttp2_session_set_local_window_size(session, NGHTTP2_FLAG_NONE, 0, 1 << 30);
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H2Connection.cpp]:
```suggestion
  nghttp2_session_set_local_window_size(session, NGHTTP2_FLAG_NONE, 0, window_size);
```

File: 3rdParty/fuerte/src/H2Connection.cpp [hunk 8]
```diff
@@ -0,0 +1,808 @@
...
+  req->append("Upgrade: h2c\r\nHTTP2-Settings: ");
+  req->append(encoded);
+  req->append("\r\n\r\n");
+
+  std::cout << "sending request '" << *req << "'\n";
+
+  auto self = Connection::shared_from_this();
+  asio_ns::async_write(
+      this->_proto.socket, asio_ns::buffer(req->data(), req->size()),
+      [self](asio_ns::error_code const& ec, std::size_t nsend) {
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/H2Connection.cpp]:
```suggestion
      [self = Connection::shared_from_this()](asio_ns::error_code const& ec, std::size_t nsend) {
```

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/8193/

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/8200/

**Pull request merged**